feat : 홈화면 추천 스터디

junggyo1020 opened the pull request:
## 📌 Related Issue
<!-- issue 링크 -->
BE-44

## 🚀 Description
<!-- 작업 내용 -->
홈 화면에 표시될 스터디 추천 기능 API를 구현했습니다.

사용자에게 3가지 유형의 추천 스터디를 각각 1개씩 보여주도록 구현했습니다.
- 이번주 가장 많이 활동한 스터디 `MOST_ACTIVE_THIS_WEEK`
- 최근 가입률이 높은 스터디 `HIGH_JOIN_RATE_RECENT`
- 난이도가 유사한 스터디 `SIMILAR_DIFFICULTY`: 유저가 최근 한달 간 푼 문제 난이도랑 근접한 문제를 최근 한달 간 푼 스터디

최근 한달 간 푼 문제의 난이도, 이번 주 스터디 활동지수, 최근 가입률은 변동성이 많은 변수들이라 어떻게 관리해야 할지 고민이 많았습니다.

그래서 저는 사전에 해당 지수들을 기준을 만들어 모두 집계해 점수로 만들었고, 스냅샷 형식으로 저장해 API 성능을 개선하고자 했습니다.

자세한 고민 사항들은 노션 `SERVER` > `개인페이지`에 기록해 두었으니 참고하시면 좋을 것 같습니다!

## 📢 Review Point
<!-- 코드리뷰 할 때 더 집중해서? 봐주면 좋을 포인트들 -->
### 1. 사전에 집계하기 & 스냅샷으로 관리하기
API 요청 시마다 복잡한 계산을 수행하는 대신, 배치 작업으로 주기적으로 데이터를 집계하고 스냅샷을 생성하는 방식을 생각해봤는데요. 이렇게 하면 API는 집계된 테이블을 단순 조회하기만 하면 되서 응답 속도에 문제가 없을 것이라 생각했습니다.

아래는 이번에 추가된 집계 테이블과 그 관계입니다. StudyGroup과 User를 기반으로 각종 활동 지표를 집계하여 저장하는 구조입니다.

```
    STUDY_GROUP 
      |-- GROUP_ACTIVITY_WEEKLY : "주간 활동 집계"
      |-- GROUP_JOIN_MONTHLY_ROLLING : "월간 가입률 집계"
      |-- GROUP_DIFFICULTY_MONTHLY_ROLLING : "스터디 난이도 집계"
      |-- STUDY_GROUP_TAG : "태그 스냅샷"
    
    USER - USER_DIFFICULTY_MONTHLY_ROLLING : "사용자 난이도 집계"
```

### 2. 서비스 로직에 대하여
각각의 서비스 로직에는 제가 임의로 알고리즘과 기준을 정의해놓았는데, 해당 부분이 괜찮을지 확인해주시면 감사하겠습니다 :)
- `MOST_ACTIVE_THIS_WEEK`, `HIGH_JOIN_RATE_RECENT` 태그는 `StudyGroupTag` 스냅샷 테이블에서 점수가 가장 높은 항목을 조회합니다 -> buildFromTagSnapshot
- `SIMILAR_DIFFICULTY `태그는 사용자의 최근 30일 평균 난이도와 가장 차이가 적은 스터디를 찾습니다. -> buildSimilarDifficulty
- 추천 데이터가 없는 경우, 가장 최근에 생성된 스터디를 반환하도록 Fallback 로직을 추가했습니다. -> fallbackFromRecentGroup
 
개선할 수 있는 부분이 있다면 많은 의견 부탁드립니다..!

### 3. Native Query
사용자와 가장 유사한 난이도의 스터디를 효율적으로 찾기 위해 **Native Query**를 사용했습니다.

ABS(g.avg_difficulty - :userAvg)로 난이도 차이를 절댓값을 기준으로 정렬한 다음, LIMIT 1로 가장 유사한 스터디를 조회하도록 했습니다. 성능상으로 더 나은 방식이 있다면 조언 부탁드리겠습니다 :)

```Java
// GroupDifficultyMonthlyRollingRepository.java
@Query(value = """
        SELECT *
        FROM group_difficulty_monthly_rolling g
        WHERE g.window_start = :windowStart AND g.window_end = :windowEnd
        ORDER BY ABS(g.avg_difficulty - :userAvg) ASC
        LIMIT 1
        """, nativeQuery = true)
Optional<GroupDifficultyMonthlyRolling> findTopSimilarByWindow(...);
```

## 📚Etc (선택)
<!-- 작업 도중 생긴 특이사항, 또는 고려할 것들(?) -->
### 테스트 결과

 **`RecommendationControllerTest`**
<img width="690" height="184" alt="image" src="https://github.com/user-attachments/assets/e4a6dcd7-ae61-47e7-94a8-2e676cc3972a" />

**`RecommendationServiceTest`**
<img width="710" height="297" alt="image" src="https://github.com/user-attachments/assets/164756e0-27fa-451b-9b4c-f2bd6c53f416" />

**`RecommendationSchedulerTest`**
<img width="779" height="250" alt="image" src="https://github.com/user-attachments/assets/874b47b8-8771-482d-adf3-bfbc722cab7a" />

**`RecommendationBatchServiceTest`**
<img width="724" height="273" alt="image" src="https://github.com/user-attachments/assets/125e179e-103c-4730-8cbb-a5f6aca727c2" />


### Swagger 테스트 결과
<img width="705" height="235" alt="image" src="https://github.com/user-attachments/assets/fd5e6363-81f8-4e5f-8454-603650866f62" />
<img width="696" height="393" alt="image" src="https://github.com/user-attachments/assets/eb62bd1b-0004-4040-bb4f-84d1c64cb907" />
<img width="699" height="369" alt="image" src="https://github.com/user-attachments/assets/fcc0df4f-4b35-4d21-9ffc-c2d6b7810ffe" />

junggyo1020 committed:
[Feat] 도메인 모델 및 TagType 정의
: 추천 스터디의 종류를 나타내는 TagType Enum 클래스 추가
- MOST_ACTIVE_THIS_WEEK: 이번주 가장 많이 활동한 스터디
- HIGH_JOIN_RATE_RECENT: 최근 가입률이 높은 스터디
- SIMILAR_DIFFICULTY: 난이도가 유사한 스터디

junggyo1020 committed:
[Feat] JPA 엔티티 추가
- GroupActivityWeekly: 주간 활동 집계
- GroupJoinMonthlyRolling: 월간 가입률 집계
- GroupDifficultyMonthlyRolling: 스터디 월간 난이도 집계
- UserDifficultyMonthlyRolling: 사용자 월간 난이도 집계
- StudyGroupTag: 태그별 기록 저장

junggyo1020 committed:
[Feat] 추천 스터디 관련 repository 추가
- StudyGroupTagRepository: 태그별 Top 1 조회 쿼리 메서드 정의
- UserDifficultyMonthlyRollingRepository: 사용자의 최근 난이도 조회 쿼리 메서드 정의
- GroupDifficultyMonthlyRollingRepository: 사용자와 가장 유사한 난이도의 스터디를 찾는 Native Query 정의

junggyo1020 committed:
[Feat] 홈 추천 스터디 API 응답 DTO 정의
- StudyGroupSummaryDto: 스터디 그룹의 요약 정보
- RecommendationItemDto: 태그별 추천 항목 정보
- HomeRecommendationsResponse: 최종 응답 DTO

junggyo1020 committed:
[Feat] 홈 추천 스터디 API 비즈니스 로직 구현
- 태그 스냅샷 테이블(StudyGroupTag)에서 '가장 활발한 스터디', '가입률 높은 스터디' 조회
- 사용자와 스터디의 최근 30일 평균 난이도를 비교하여 '유사 난이도 스터디' 조회
- 각 항목 조회 실패 시, 가장 최근에 생성된 스터디를 반환하는 Fallback 로직 추가

junggyo1020 committed:
[Feat] 홈 추천 스터디 조회 API 컨트롤러 구현

junggyo1020 committed:
[Test] 추천 기능 서비스 & 컨트롤러 테스트 코드 작성
- RecommendationServiceTest: 추천 스터디 관련 시나리오에 대한 단위 테스트
- RecommendationControllerTest: API 정상 응답 확인을 위한 MockMvc 테스트

junggyo1020 committed:
[Feat] 추천 스터디 집계 배치 서비스 구현

- RecommendationBatchService: 스케줄링될 추천 스터디 집계 로직 구현
- calculateAndSaveMostActiveStudy: 주간 활동 점수 계산 및 스냅샷 저장
- calculateAndSaveHighJoinRateStudy: 월간 가입률 계산 및 스냅샷 저장
- calculateAndSaveDifficultyInfo: 스터디/사용자 난이도 데이터 집계
- Solution/Problem/SolutionCommentRepository: 집계에 필요한 JPQL 쿼리 메서드 추가

junggyo1020 committed:
[Feat] 스케쥴러 작성

junggyo1020 committed:
[Test] 추천 기능 서비스 & 컨트롤러 테스트 코드 작성

- RecommendationServiceTest: 추천 스터디 관련 단위 테스트
- RecommendationControllerTest: API 정상 응답 확인을 위한 MockMvc 테스트

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/controller/RecommendationController.java:
```diff
@@ -0,0 +1,35 @@
+	@Operation(summary = "홈 추천 스터디 조회 API")
+	public ResponseEntity<HomeRecommendationsResponse> getHomeRecommendations(
+		@AuthedUser User user,
+		@RequestParam(name = "userId", required = false) Long userId
```
아마 파라미터로 받은 userId가 현재 로그인한 회원을 의미하신 것 같아요!
`@AuthedUser`로 받은 User 객체가 인증 정보를 파싱해서 가져온 회원 정보입니다.
(`AuthedUserResolver` 보시면 jwt 토큰 값을 사용해서 유저 정보를 불러오고 있어요!)
홈화면은 개인화된 영역이니까 request param 없이 `User` 넘겨서 사용해도 괜찮을 것 같아요!

junggyo1020 replied:
헉 @AuthedUser가 이미 인증된 사용자 정보를 안전하게 가져오고 있네요! 제가 이부분을 놓친 것 같아요..! 굳이 userId를 쿼리 파라미터로 따로 받을 필요가 없으니 이 부분은 삭제하도록 하겠습니다! 좋은 의견 주셔서 정말 감사합니다 :)

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/GroupActivityWeekly.java:
```diff
@@ -0,0 +1,51 @@
+@Entity
+@Table(name = "group_activity_weekly")
+@Getter
+@NoArgsConstructor
```
외부 코드에서 직접적으로 잘못 생성하는 것을 방지하기 위해 `@NoArgsConstructor(access = AccessLevel.PROTECTED)` 설정 해주는 것도 좋을 것 같아요!

junggyo1020 replied:
네 이 부분 추가해 두겠습니다 👍

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/GroupActivityWeekly.java:
```diff
@@ -0,0 +1,51 @@
+@Table(name = "group_activity_weekly")
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
지금 다른 엔티티들에는 추가가 안되어있긴한데.. 아래처럼 constructor에 검증을 추가하는건 어떨까요??
전에 builder 쓰다가 필드를 하나씩 빼먹는 실수가 생겼었는데, 아래처럼 검증 추가하니 좀 낫더라구요..ㅎㅎ

```suggestion
@Builder
public GroupActivityWeekly(
        Long id,
        @NotNull StudyGroup studyGroup,
        @NotNull LocalDateTime weekStart,
        @NotNull LocalDateTime weekEnd,
        @NotNull Integer submissions,
        @NotNull Integer comments,
        @NotNull Double activeScore
) {
    this.id = id;
    this.studyGroup = studyGroup;
    this.weekStart = weekStart;
    this.weekEnd = weekEnd;
    this.submissions = submissions;
    this.comments = comments;
    this.activeScore = activeScore;
}
```

부가적인 부분이라 적용은 안하셔도 괜찮습니다!

junggyo1020 replied:
null 값이 들어가거나 나중에 save() 단계에서 NPE가 터질 수도 있겠네요 ㅎㅎ

생성자 레벨에서 @Builder를 사용하고 @NotNull로 명확하게 검증 로직을 추가하는 방식이 훨씬 더 안정적이고 견고한 코드가 될 것 같아서 GroupActivityWeekly 와 GroupJoinMonthlyRolling 등 다른 엔티티들에도 동일하게 적용해두겠습니다 :)

좋은 제안 감사드립니다 :)

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/StudyGroupTag.java:
```diff
@@ -0,0 +1,57 @@
+	private StudyGroup studyGroup;
+
+	@Enumerated(EnumType.STRING)
+	@Column(name = "tag_type", nullable = false, length = 64)
```
꼼꼼한 length 처리 좋습니다 👍

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/GroupActivityWeekly.java:
여기 엔티티에 남긴 코멘트들은 추가하신 다른 모든 엔티티들에도 해당하는 공통 코멘트입니다!
반영하시게 된다면 여기 도메인의 타 엔티티들에도 같이 적용해주시면 좋을 것 같습니다 :)

junggyo1020 replied:
확인했습니다..! 적용은 모두 해두겠습니다 :)

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/dto/StudyGroupSummaryDto.java:
```diff
@@ -0,0 +1,11 @@
+        String name,
+        String introduction,
+        String groupImage,
+        List<String> tags
```
`RecommendationItemDto`에 `TagType`이 있어서 여기에 있는 `List<String> tags`는 필요 없을 것 같은데,
혹시 여기에도 두신 이유가 따로 있는걸까요??

junggyo1020 replied:
아래의 리스트 형식의 태그는 하나의 스터디 그룹이 여러 개의 태그를 가질 수가 있어서, 어떤 추천 스터디인지가 위의 태그타입으로 분류되고, 아래의 태그는 해당 스터디가 가진 모든 태그들을 불러오도록 구현했습니다 :)

rladmstn replied:
아아 추천할 때 해당 그룹의 모든 태그를 보여주는군요!
확인했습니다!

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/repository/GroupActivityWeeklyRepository.java:
```diff
@@ -0,0 +1,8 @@
+import com.gamzabat.algohub.feature.recommendation.domain.GroupActivityWeekly;
+
+public interface GroupActivityWeeklyRepository extends JpaRepository<GroupActivityWeekly, Long> {
+}
```
마지막 줄 개행 추가 부탁드립니다!

junggyo1020 replied:
넵! 확인했습니다. 개행 문자 추가해서 반영할게요! 😃

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/repository/GroupDifficultyMonthlyRollingRepository.java:
```diff
@@ -0,0 +1,24 @@
+
+public interface GroupDifficultyMonthlyRollingRepository extends JpaRepository<GroupDifficultyMonthlyRolling, Long> {
+
+    @Query(value = """
+            SELECT *
+            FROM group_difficulty_monthly_rolling g
+            WHERE g.window_start = :windowStart AND g.window_end = :windowEnd
+            ORDER BY ABS(g.avg_difficulty - :userAvg) ASC
+            LIMIT 1
+            """, nativeQuery = true)
```
최대한 성능 측면을 고려하면 지금 상태에서는 `ORDER BY ABS(g.avg_difficulty - :userAvg) ASC` 부분이 인덱스 활용을 못하게 돼서 가장 부하 포인트가 될 것 같아보이네요!

`userAvg` 값보다 큰 것 들 중 최솟값, `userAvg` 값보다 작은 것 중 최댓값 두 개를 구해서 그 둘의 절댓값 중 더 작은 걸 반환하도록 만드는건 어떨까 싶습니다. 인덱스를 활용하기엔 제일 좋은 방법이 될 것 같아보여서요!
아래처럼 사용하면 `ABS()` 연산도 최대 두 개 데이터에 대해서만 하게 될 것 같아요.

```sql
SELECT val
FROM (
    -- userAvg보다 큰 값 중 최소값
    SELECT MIN(avg_difficulty) AS val
    FROM group_difficulty_monthly_rolling g
    WHERE g.window_start = :windowStart
      AND g.window_end = :windowEnd
      AND g.avg_difficulty >= :userAvg

    UNION ALL

    -- userAvg보다 작은 값 중 최대값
    SELECT MAX(avg_difficulty) AS val
    FROM group_difficulty_monthly_rolling
    WHERE window_start = :windowStart
      AND window_end = :windowEnd
      AND avg_difficulty <= :userAvg
) t
ORDER BY ABS(val - :userAvg)
LIMIT 1;
```

물론 쿼리가 복잡해져서.. 미리 성능 개선하기 vs 가독성 유지하기 가 될 것 같긴 합니다!
선택사항이 될 것 같아요!

junggyo1020 replied:
이 부분 제가 실제로 성능 테스트 진행해보면서 하나의 PR로 별도로 분리를 해보고 싶네요 ㅎㅎ
좋은 의견 감사합니다 :) 일단 이 PR에서는 반영하지 않고 넘기겠습니다!!!

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/repository/GroupJoinMonthlyRollingRepository.java:
```diff
@@ -0,0 +1,8 @@
+import com.gamzabat.algohub.feature.recommendation.domain.GroupJoinMonthlyRolling;
+
+public interface GroupJoinMonthlyRollingRepository extends JpaRepository<GroupJoinMonthlyRolling, Long> {
+}
```
마지막 줄 개행 추가 부탁드립니다!

junggyo1020 replied:
넵! 확인했습니다. 개행 문자 추가해서 반영할게요! 😃

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/UserDifficultyMonthlyRolling.java:
```diff
@@ -0,0 +1,39 @@
+	private Long id;
+
+	@Column(name = "user_id", nullable = false)
+	private Long userId;
```
여기 `User` 엔티티 사용해서 연관관계 매핑 해줘도 될 것 같아요!

junggyo1020 replied:
생각해보니 말씀대로 @ManyToOne을 사용해서 User 엔티티와 직접 연관관계를 맺어주는 것이 훨씬 더 객체지향적인 것 같네요!

User가 삭제될 경우에 데이터 무결성을 관리하는 측면이나, 나중에 집계 데이터를 활용할 때 User 객체로 바로 탐색할 수 있으니 확장성 측면에서도 훨씬 유리할 것 같아요.

좋은 의견 감사합니다..! 이 수정은 Repository와 Service 코드에도 연쇄적으로 변경이 필요해서, 꼼꼼하게 확인 후 바로 수정 반영하겠습니다!

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	/**
+	 * 매일 오전 4시 10분 -> '최근 가입률이 높은 스터디' 집계
```
활동량 적을 때 돌리는 스케줄러 좋습니다 👍

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/scheduler/RecommendationScheduler.java:
```diff
@@ -0,0 +1,42 @@
+	public void scheduleDifficultyInfo() {
+		recommendationBatchService.calculateAndSaveDifficultyInfo();
+	}
+}
```
여기도 개행 추가 부탁합니다!

junggyo1020 replied:
넵! 확인했습니다. 개행 문자 추가해서 반영할게요! 😃

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/service/RecommendationBatchService.java:
```diff
@@ -0,0 +1,210 @@
+				.build());
+		});
+	}
+}
```
여기도 부탁드립니다 ㅎㅎ

junggyo1020 replied:
넵! 확인했습니다. 개행 문자 추가해서 반영할게요! 😃

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/service/RecommendationBatchService.java:
```diff
@@ -0,0 +1,210 @@
+					weekEnd);
+				Long comments = solutionCommentRepository.countByStudyGroupAndCreatedAtBetween(group, weekStart,
+					weekEnd);
+				double activeScore = (submissions * 1.0) + (comments * 1.5);
```
코멘트에 대한 가중치 좋아요! 👍

rladmstn commented on src/main/java/com/gamzabat/algohub/feature/recommendation/domain/GroupActivityWeekly.java:
```diff
@@ -0,0 +1,51 @@
+	@JoinColumn(name = "study_group_id", nullable = false)
+	private StudyGroup studyGroup;
+
+	@Column(name = "week_start", nullable = false)
+	private LocalDateTime weekStart;
+
+	@Column(name = "week_end", nullable = false)
+	private LocalDateTime weekEnd;
```
특정 시간이 중요한게 아니라면 `LocalDate`로 써도 괜찮을 것 같아보여요!
아님 혹시 Time까지 필요한 경우가 따로 있을까요??

junggyo1020 replied:
이렇게 세세한 부분까지 짚어주셔서 정말 감사합니다. 
만약 "날짜" 기준으로만 집계한다면 LocalDate를 쓰는 것이 훨씬 좋은 것 같다고 생각해요.

제가 LocalDateTime을 사용한 이유는, 이 GroupActivityWeekly가 집계하는 대상 컬럼들인 Solution의 `solvedDateTime`과 SolutionComment의 `createdAt`이 모두 `LocalDateTime` 타입이기 때문이에요.

weekStart = 지난주 월요일 00:00:00 (atStartOfDay())

weekEnd = 지난주 일요일 23:59:59... (plusWeeks(1).minusNanos(1))
이렇게 정확한 시간 경계를 계산하게 되는데요!

"지난주"라는 기간을 정확하게 집계하기 위해 GroupActivityWeekly 테이블의 weekStart, weekEnd 컬럼은 `LocalDateTime`으로 설계했습니다.

만약 이 컬럼을 LocalDate (예: 2025-10-13 ~ 2025-10-19)로 저장하면, 이 데이터가 13일 00시부터 19일 23시 59분까지를 의미하는지, 아니면 13일 00시부터 19일 00시(20일 00시 전)까지를 의미하는지 모호해질 수 있다고 판단했습니다. `LocalDateTime`으로 저장함으로써 이 activeScore가 어떤 시간 범위(Timestamp)를 기준으로 계산되었는지 명확하게 기록하고자 했습니다.

꼼꼼한 리뷰 정말 감사합니다!

rladmstn replied:
오오 코드 내에서 로직의 가독성도 고려하셨던 판단이군요..!
이런 의도셨다면 `LocalDateTime` 쓰는게 좋은 것 같아요! 이해됐습니다 설명 감사합니다!!

rladmstn commented on src/test/java/com/gamzabat/algohub/feature/recommendation/domain/GroupActivityWeeklyTest.java:
```diff
@@ -0,0 +1,60 @@
+			.build();
+
+		// then
+		assertThat(activity.getStudyGroup()).isEqualTo(studyGroup);
+		assertThat(activity.getWeekStart()).isEqualTo(weekStart);
+		assertThat(activity.getWeekEnd()).isEqualTo(weekEnd);
+		assertThat(activity.getSubmissions()).isEqualTo(10);
+		assertThat(activity.getComments()).isEqualTo(5);
+		assertThat(activity.getActiveScore()).isEqualTo(17.5);
```
아래처럼 쓰면 실패해도 나머지 assertion은 계속 검증해서 모든 실패를 한 번에 보기 좋습니다!
반영하실 필요는 없고 혹시나 싶어서 참고용으로 코멘트 남깁니다..!

```suggestion
        Assertions.assertAll(
            () -> assertThat(joinInfo.getStudyGroup()).isEqualTo(studyGroup),
            () -> assertThat(joinInfo.getWindowStart()).isEqualTo(windowStart),
            () -> assertThat(joinInfo.getWindowEnd()).isEqualTo(windowEnd),
            () -> assertThat(joinInfo.getNewMembers()).isEqualTo(5),
            () -> assertThat(joinInfo.getMembersBeforeWindow()).isEqualTo(20),
            () -> assertThat(joinInfo.getJoinRate()).isEqualTo(0.25)
        );
```

junggyo1020 replied:
좋은 팁 너무 감사합니다!

rladmstn reviewed (commented):
테스트까지 꼼꼼하게..! 고생하셨습니다!!
도메인부터 코드 이해까지 쉽지 않으셨을텐데 엄청 잘 구현해주셨네요!
PR이랑 노션에도 고민한 부분들 세세하게 잘 남겨주셔서 생각하신 설계를 이해하기 쉬웠습니다 👍

---

제가 이해한 바로는 `GroupActivityWeekly`, `GroupJoinMonthlyRolling`, `GroupDifficultMonthlyRolling`, `UserDifficultyMonthlyRolling` 네 가지에 대해 매일 오전 4시쯤에 스케줄러로 계산해서 데이터를 insert 하는 것 같아요!

개인적으로는 각각의 데이터들이 `StudyGroup`, `User` 당 하나씩 unique한 레코드로 존재해도 되지 않을까 싶은데요, 그래서 처음 생성된 데이터를 가지고 계속 update를 진행해도 괜찮을 것 같았습니다!
update가 아닌 insert로 데이터를 갱신하신 이유가 무엇인지 궁금합니다!
(혹시 notion에 있는데 제가 놓친거라면.. 보고 오겠습니다 😅)

junggyo1020 committed:
[Refactor] AuthedUser를 통해 User 객체를 받아와 Param으로 사용하도록 수정

junggyo1020 committed:
[Refactor] Builder와 함께 NotNull 검증 로직 추가

junggyo1020 committed:
[Test] 테스트 코드 수정

junggyo1020 committed:
[Refactor] 개행 추가

junggyo1020 committed:
[Refactor] 개행 추가

junggyo1020 committed:
[Refactor] 개행 추가

junggyo1020 committed:
[Refactor] User 엔티티와 직접 @ManyToOne 연관관계 맺도록 수정

junggyo1020 committed:
[Test] 테스트 코드 수정

junggyo1020 committed:
[Test] Assertions.assertAll 적용
- 실패해도 나머지 assertion은 계속 검증해서 모든 실패를 한 번에 보기 좋다는 의견을 참고해 반영했습니다 :)

junggyo1020 commented:
추가적으로 Update가 아닌 Insert로 구현한 이유에 대해 말씀 드리려고 해요.

말씀해주신 것처럼 StudyGroup이나 User별로 유니크한 레코드를 두고 UPDATE 하는 방식도 충분히 고려할 수 있는, 그리고 성능 면에서도 장점이 있는 좋은 접근 방식이라고 생각합니다. 실제로 두 가지 방식을 두고 고민했었는데요, 아래와 같은 이유로 INSERT 방식을 최종적으로 선택하게 되었습니다.

1. 데이터 보존과 추이 분석을 위해서
: 매번 새로운 레코드로 집계 데이터를 쌓게 되면, 특정 스터디의 과거 활동량 변화나 가입률 추이와 같은 시계열 데이터를 쉽게 분석할 수 있습니다. 예를 들어, "A 스터디가 지난달에 비해 이번 달에 얼마나 더 활발해졌는가?"와 같은 지표를 추출하거나, 나중에 사용자에게 "우리 스터디 활동량 그래프" 같은 기능을 제공할 때 매우 유용하다고 판단했습니다. `UPDATE` 방식은 항상 최신 상태만 덮어쓰기 때문에 이러한 추이 분석이 불가능하다는 단점이 있다고 생각했고, **기획의 방향성에 따라 이 부분은 달라질 수 있을 것이라 판단했습니다 :)**

2. 데이터 무결성 및 디버깅 용이성
: INSERT 방식은 특정 시점의 집계 결과를 '스냅샷'처럼 그대로 보존하기 때문에 데이터가 변경될 여지가 없고, 만약 특정 날짜의 추천 결과에 문제가 생겼을 때, 당시 집계 데이터를 그대로 조회하여 원인을 파악하기 용이하다는 장점을 생각했습니다.

위 두가지 이유로 당장의 저장 공간 비용이나 UPDATE로 구현할 때 단순함보다 더 중요하다고 생각했고, 데이터를 보존하는 `INSERT` 방식으로 결정하게 되었습니다.

다른 분들의 생각은 어떠신지, 혹시나 제가 놓치고 있는 부분이 있다면 편하게 말씀해주세요! 다시 한번 좋은 질문 감사드립니다. 👍

junggyo1020 committed:
[Feat] 홈 화면 추천 API 응답에 스터디 기간 추가

- 클라이언트 요청에 따라 StudyGroupSummaryDto에 startDate, endDate 필드 추가
- RecommendationService의 DTO 변환 로직(toSummary)에 해당 필드 매핑
- DTO 변경으로 인한 RecommendationControllerTest, RecommendationServiceTest 수정

rladmstn reviewed (approved):
현재 기획 방향성이 유동적일 수 있다는 점 고려해서 insert로 구현하신 점 공감합니다!
저도 현재의 데이터들이 내부적인 분석 뿐 만 아니라, 더 다양한 방법으로 사용자들에게 보여줄 수 있을 것 같아서 현재의 정교님 구현 방식에 동의합니다 :)
설명 감사합니다! 

코멘트 반영 잘해주셔서 추가 피드백은 없는 것 같아요. 고생하셨습니다 -

sh0723 commented:
정교님 말대로 update 보다 insert가 데이터보존 및 추이분석을 위해 더 맞는 선택인거같아요! 
설명을 세세하게 잘 써놔주셔서 이해하는데 도움이 많이 됐습니다!
고생하셨어요~!

s-hwan reviewed (approved):
배운다는 생각으로 읽어봤습니다 노션까지 생각을 어떤식으로 하셨는지 알 수 있게 정리해주셔서 감사합니다 👍

junggyo1020 committed:
[Feat] 추천 스터디 집계 관련 메서드 추가
